File: 卡码/代码随想录/1/07-回溯/Python/06-其他/LeetCode_0491.py
#491. 非递减子序列
from typing import *

class Solution:
  def __init__(self):
    self.path:List[int]=[]
    self.res:List[List[int]]=[]
  def backtracking(self,nums:List[int],index:int):
    if len(self.path)>1:
      self.res.append(self.path[:])
    hash:List[bool]=[False] * 201
    for i in range(index,len(nums)):
      if len(self.path)>0 and nums[i]<self.path[-1] or hash[nums[i]+100]:
        continue
      hash[nums[i]+100]=True
      self.path.append(nums[i])
      self.backtracking(nums,i+1)
      self.path.pop()
  def findSubsequences(self, nums: List[int]) -> List[List[int]]:
    self.backtracking(nums,0)
    return self.res

File: 卡码/代码随想录/1/07-回溯/Js/06-其他/LeetCode_0491.js
//491. 非递减子序列
const findSubsequences = function(nums) {
  const res=new Array();
  const path=new Array();
  const backtracking=function(nums,index){
    if(path.length>1) res.push(Array.from(path));
    const hash=new Array(201).fill(false);
    for(let i=index;i<nums.length;i++){
      if(path.length>0&&nums[i]<path[path.length-1]||hash[nums[i]+100]) continue;
      hash[nums[i]+100]=true;
      path.push(nums[i]);
      backtracking(nums,i+1);
      path.pop();
    }
  }
  backtracking(nums,0);
  return res;
};